### PR TITLE
Fix #1388 -- Prevent some words from occurring in order codes

### LINKED ISSUE #1388: Filtering order numbers for offensive letter/number combinations
Just had an order with the order number "FUCK9", which is quite funny, if somebody belonging to your own team orders tickets. But I guess, that some customers are perhaps not quite so humorous. 😆 

So eventually the "order number generator" could go through a list of offensive letter/number combinations (like "FUCK#", "SEX##") etc. to avoid such 'words'.

### PR DESCRIPTION
**Pull request by Sohalt**
A very rudimentary implementation to prevent some offensive substrings in order codes.
Not sure if you want to merge this in the current state. Possible concerns/improvements:
- allow to provide custom blacklist/disable blacklist
- allow to provide different blacklists per locale
- more "intelligent" filtering, respecting word boundaries (currently it only looks for substrings, so an order code of "classic" would be marked offensive, because it contains the substring "ass")

 Fixes #1388

### CODE REVIEW

**Comment by raphaelm** [comment]
Do we also want this for auto-generated voucher codes?

> allow to provide custom blacklist/disable blacklist

No, I do not think we should make this configurable.

> allow to provide different blacklists per locale

Hmm. There is a resource here:
https://github.com/LDNOOBW/List-of-Dirty-Naughty-Obscene-and-Otherwise-Bad-Words
It's under CC license though, which might make using it more complicated.

If we add much more words, I fear a performance impact. We should measure that in some way and check if the loop is really the fastest implementation. A pre-compiled, combined regex ``(sex|fuck|drugs|…)`` *might* even be faster – can you check?

> more "intelligent" filtering, respecting word boundaries (currently it only looks for substrings, so an order code of "classic" would be marked offensive, because it contains the substring "ass")

I think that is fine, given the randomness and shortness of order codes.

**Comment by Sohalt** [comment]
> Do we also want this for auto-generated voucher codes?

If we want this for voucher codes we should take into account the prefix. We then would need to reject prefixes which already contain blacklisted words, because otherwise we would have an infinite loop (because the prefix never changes). This would need to be communicated in the frontend when setting a prefix.

> Hmm. There is a resource here:
> https://github.com/LDNOOBW/List-of-Dirty-Naughty-Obscene-and-Otherwise-Bad-Words
> It's under CC license though, which might make using it more complicated.

I guess since it is only CC-BY there should be no problem using it if we add a comment stating where it is from. It might be nice to also have an "about" page or similar somewhere within pretix, listing dependencies and their licenses, where we then could also add this.

> If we add much more words, I fear a performance impact. We should measure that in some way and check if the loop is really the fastest implementation. A pre-compiled, combined regex `(sex|fuck|drugs|…)` _might_ even be faster – can you check?

I've briefly tested the loop, a pre-compiled regex, and a pre-compiled regex with `re.IGNORECASE`. The pre-compiled regex without `re.IGNORECASE` was fastest.

> I think that is fine, given the randomness and shortness of order codes.

What about vouchers though?

**Comment by raphaelm** [comment]
> If we want this for voucher codes we should take into account the prefix. We then would need to reject prefixes which already contain blacklisted words, because otherwise we would have an infinite loop (because the prefix never changes). This would need to be communicated in the frontend when setting a prefix.

I don't think we should prevent people from using specific prefixes. How about just checking the random part?

```
def _generate_random_code(prefix=None):
    charset = list('ABCDEFGHKLMNPQRSTUVWXYZ23456789')
    rnd = None
    while not rnd or blacklisted(rnd):
        rnd = get_random_string(length=settings.ENTROPY['voucher_code'], allowed_chars=charset)
    if prefix:
        return prefix + rnd
    return rnd
```

> I guess since it is only CC-BY there should be no problem using it if we add a comment stating where it is from. It might be nice to also have an "about" page or similar somewhere within pretix, listing dependencies and their licenses, where we then could also add this.

Hmm, I guess. Adding a license page like this is not an easy task, though ;)

> I've briefly tested the loop, a pre-compiled regex, and a pre-compiled regex with re.IGNORECASE. The pre-compiled regex without re.IGNORECASE was fastest.

That's what I expected, thanks for checking!

> What about vouchers though?

Right, those are longer, but we're still only checking on random things. If we really reject a string containing "classic" because of the "ass", what's the harm? It just costs us another loop iteration.

**Comment by Sohalt** [comment]
> I don't think we should prevent people from using specific prefixes. How about just checking the random part?
> 
> ```
> def _generate_random_code(prefix=None):
>     charset = list('ABCDEFGHKLMNPQRSTUVWXYZ23456789')
>     rnd = None
>     while not rnd or blacklisted(rnd):
>         rnd = get_random_string(length=settings.ENTROPY['voucher_code'], allowed_chars=charset)
>     if prefix:
>         return prefix + rnd
>     return rnd
> ```

This can lead to situations where the random part itself does not contain any offensive words, but the concatenation of prefix+rnd does. But I guess it would be a trade-off worth making.

> Hmm, I guess. Adding a license page like this is not an easy task, though ;)

True. But it's probably even enough to just mention the source in a comment.
 
> Right, those are longer, but we're still only checking on random things. If we really reject a string containing "classic" because of the "ass", what's the harm? It just costs us another loop iteration.

The longer the string the higher the probability that one of the blacklisted words shows up in there somewhere. Especially with short blacklisted words like "ass". The question would be if we only want to block words which are surrounded by a "boundary", i.e. a number and accept strings like "fooassbar", since the blacklisted word is not as apparent in there compared to "foo1ass2bar". But I guess the probabilities should still be low enough that it doesn't matter.

**Comment by raphaelm** [comment]
> This can lead to situations where the random part itself does not contain any offensive words, but the concatenation of prefix+rnd does. But I guess it would be a trade-off worth making.

I think so ;)

> True. But it's probably even enough to just mention the source in a comment.

I think about it. For now, the English list of words is a good start, I guess.

> The longer the string the higher the probability that one of the blacklisted words shows up in there somewhere. Especially with short blacklisted words like "ass". The question would be if we only want to block words which are surrounded by a "boundary", i.e. a number and accept strings like "fooassbar", since the blacklisted word is not as apparent in there compared to "foo1ass2bar". But I guess the probabilities should still be low enough that it doesn't matter.

Yes, the probabilities are higher, but the regex to match these boundaries will be mich slower, so I guess we can live with it. It would be intereseting to see how much this tunes down the entropy of voucher codes and especially order codes, though :D

**Comment by Sohalt** [comment]
> Yes, the probabilities are higher, but the regex to match these boundaries will be mich slower, so I guess we can live with it. It would be intereseting to see how much this tunes down the entropy of voucher codes and especially order codes, though :D

I'm not sure how python's regex library is implemented internally, but matching something like "[0-9]foo[0-9]" should be doable using DFAs, which would be linear in the length of the input string, i.e. pretty fast.

File: src/pretix/base/blacklist.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import re
+
+# blacklist based on http://www.bannedwordlist.com/lists/swearWords.txt
+blacklist = [
```
[the remark this answers is not in the document]
Comment (Sohalt) [reply src/pretix/base/blacklist.py]: Good point, I've adjusted the code. While doing so I found a couple of other occurrences of "blacklist" in the code. I've also replaced those and opened  #1434 for that.

**Pull request merged**